File: scripts/smoke-test-mcp.mjs
#!/usr/bin/env node

import { spawn } from "node:child_process";

const [command, ...args] = process.argv.slice(2);
if (!command) {
  process.stderr.write("Usage: scripts/smoke-test-mcp.mjs <command> [args...]\n");
  process.exit(2);
}

const child = spawn(command, args, { stdio: ["pipe", "pipe", "pipe"] });
let pendingStdout = "";
let stderr = "";
let completed = false;

const timeout = setTimeout(() => finish(new Error(`MCP smoke test timed out. stderr:\n${stderr}`)), 10_000);
child.stdout.setEncoding("utf8");
child.stderr.setEncoding("utf8");
child.stderr.on("data", chunk => {
  stderr += chunk;
});
child.stdout.on("data", chunk => {
  pendingStdout += chunk;
  const lines = pendingStdout.split("\n");
  pendingStdout = lines.pop() ?? "";
  for (const line of lines.filter(Boolean)) {
    let message;
    try {
      message = JSON.parse(line);
    } catch {
      finish(new Error(`MCP stdout contained non-JSON data: ${line}`));
      return;
    }
    if (message.id === 1 && message.result) {
      child.stdin.write('{"jsonrpc":"2.0","method":"notifications/initialized"}\n');
      child.stdin.write('{"jsonrpc":"2.0","id":2,"method":"tools/list","params":{}}\n');
    }
    if (message.id === 2 && Array.isArray(message.result?.tools)) {
      const names = new Set(message.result.tools.map(tool => tool.name));
      if (!names.has("jaipilot_prepare_tests") || !names.has("jaipilot_prepare_cleanup")) {
        finish(new Error(`MCP tool list is incomplete: ${[...names].join(", ")}`));
        return;
      }
      finish();
      return;
    }
  }
});
child.once("error", finish);
child.once("exit", (status, signal) => {
  if (!completed) {
    finish(new Error(`MCP server exited early (${signal ?? status}). stderr:\n${stderr}`));
  }
});

child.stdin.write('{"jsonrpc":"2.0","id":1,"method":"initialize","params":{"protocolVersion":"2025-11-25","capabilities":{},"clientInfo":{"name":"smoke-test","version":"1"}}}\n');

function finish(error) {
  if (completed) return;
  completed = true;
  clearTimeout(timeout);
  child.kill("SIGTERM");
  if (error) {
    process.stderr.write(`${error.message}\n`);
    process.exitCode = 1;
  } else {
    process.stdout.write("MCP initialize and tool discovery passed\n");
  }
}

File: scripts/smoke-test-mcp.py
#!/usr/bin/env python3
"""Black-box initialize + tools/list smoke for JAIPilot's stdio MCP server."""

from __future__ import annotations

import json
import os
import selectors
import subprocess
import sys
import time


EXPECTED_TOOLS = {
    "jaipilot_inspect",
    "jaipilot_snapshot",
    "jaipilot_quality",
    "jaipilot_rewrite",
    "jaipilot_diff_gate",
    "jaipilot_prove_diff",
}


def fail(message: str) -> None:
    raise SystemExit(f"MCP smoke failed: {message}")


def response(process: subprocess.Popen[bytes], request_id: int, timeout: float = 10.0) -> dict:
    selector = selectors.DefaultSelector()
    assert process.stdout is not None
    selector.register(process.stdout, selectors.EVENT_READ)
    deadline = time.monotonic() + timeout
    while time.monotonic() < deadline:
        if process.poll() is not None:
            fail(f"server exited with {process.returncode}")
        events = selector.select(max(0.0, deadline - time.monotonic()))
        if not events:
            continue
        line = process.stdout.readline()
        if not line:
            fail("server closed stdout")
        try:
            message = json.loads(line)
        except json.JSONDecodeError as exception:
            fail(f"non-protocol stdout: {line!r} ({exception})")
        if message.get("id") == request_id:
            return message
    fail(f"timed out waiting for response {request_id}")


def send(process: subprocess.Popen[bytes], message: dict) -> None:
    assert process.stdin is not None
    process.stdin.write(json.dumps(message, separators=(",", ":")).encode() + b"\n")
    process.stdin.flush()


def main() -> None:
    if len(sys.argv) < 2:
        fail("usage: smoke-test-mcp.py <server-command> [args ...]")
    environment = os.environ.copy()
    environment.setdefault("JAIPILOT_DASHBOARD_DISABLED", "1")
    process = subprocess.Popen(
        sys.argv[1:],
        stdin=subprocess.PIPE,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        env=environment,
    )
    try:
        send(process, {
            "jsonrpc": "2.0",
            "id": 1,
            "method": "initialize",
            "params": {
                "protocolVersion": "2025-11-25",
                "capabilities": {},
                "clientInfo": {"name": "jaipilot-smoke", "version": "1"},
            },
        })
        initialized = response(process, 1)
        if "result" not in initialized or initialized.get("error") is not None:
            fail(f"initialize returned {initialized}")
        send(process, {"jsonrpc": "2.0", "method": "notifications/initialized"})
        send(process, {"jsonrpc": "2.0", "id": 2, "method": "tools/list", "params": {}})
        listed = response(process, 2)
        tools = listed.get("result", {}).get("tools", [])
        names = {tool.get("name") for tool in tools if isinstance(tool, dict)}
        if names != EXPECTED_TOOLS:
            fail(f"expected {sorted(EXPECTED_TOOLS)}, received {sorted(names)}")
        print(f"MCP initialize/tools-list passed with {len(names)} tools.")
    finally:
        process.terminate()
        try:
            process.wait(timeout=5)
        except subprocess.TimeoutExpired:
            process.kill()
            process.wait(timeout=5)


if __name__ == "__main__":
    main()
